### ISSUE TITLE
Minimum modulus size

### ISSUE THREAD

**Issue by ryancdotorg** [RustCrypto/RSA#445]
RSA is catastrophically weak with 512 bit keys, and anything less than 2048 bits is insecure. A minimum key size should be enforced. Golang has proposed to require 1024 bits minimum:

https://github.com/golang/go/issues/68762

however 2048 would be a good choice where backwards compatibility requirements are minimal

**Comment by tarcieri** [comment]
See also: #350

**Comment by dignifiedquire** [comment]
For testing and supporting existing implementations I don't think a blanket limit makes sense. 
But maybe a default generator that limits the range would be helpful for newcomers.

**Comment by ryancdotorg** [comment]
Not supporting existing implementations which would require this is a benefit.

RSA 1024 and even 2048 are plenty fast for testing.

**Comment by dignifiedquire** [comment]
> Not supporting existing implementations which would require this is a benefit.

I am glad for you that this is how you can run your projects, it is not the same for all of us.

**Comment by str4d** [comment]
I agree with @ryancdotorg and with the rationales outlined in the Go proposal. 

I also note that the `rsa` crate already enforces limits on key sizes, specifically `RsaPublicKey::new` which returns an error for any key larger than 4096 bits. This method has a trapdoor for configurability via `RsaPublicKey::new_with_max_size`.

An intermediate pathway would therefore be to do the same thing for minimum sizes:
- Have the default pathway enforce a secure minimum size in `RsaPrivateKey::new` and `RsaPublicKey::new`, which we can increase as necessary in line with best practices.
- Have a trapdoor for configurability via e.g. `RsaPrivateKey::new_with_insecure_key_size`, which enables reviewers to clearly identify codepaths that are insecure in production.

Then the question of whether to expose the insecure / hazmat configuration APIs at allcan be separated from the "stop people shooting themselves in the foot because they think they are getting 256-bit security" issue.

**Comment by ryancdotorg** [comment]
For reference, here's an example of use of 512 bit RSA that could have damaged a power grid:

https://rya.nc/vpp-hack.html

It's been known to basically worthless in terms of security for decades, let it die.

**Comment by pinkforest** [comment]
The API IMO probably should use mis-use resistance. Just like with x25519-dalek we added static_secrets opt-in to track it's use.

One way would be to split the constructor for insecure and secure constructions - the insecure constructor for given bit-lengths would be hidden behind a feature.

By having to use a feature (or cfg) to override mis-use resistance it could be potentially tracked in at least feature usage - now it is impossible to find any direct dependencies which use insecure sizes - where as using a feature it can be tracked through analysing crates.io-index that exposes the used features.

**Comment by ryancdotorg** [comment]
Extensive empirical evidence shows that users of cryptographic libraries are frequently not qualified to make choices on things like modulus size.

Putting broken (< 1024 bits) and weak (< 2048 bits) behind feature flags seems like a good solution, and type annotating by size is an interesting idea.

**Comment by tarcieri** [comment]
I'll +1 @str4d's suggestion that we should follow what we already do with `RsaPublicKey` and have checked `RsaPrivateKey` APIs with short, convenient names which enforce a minimum modulus, and parallel unchecked APIs with longer names that don't enforce the restriction

**Comment by elichai** [comment]
@tarcieri Is there any way to do this without adding to this as a fully fledged feature? Maybe some kind of `fn modulus_size(&self) -> usize` that will allow us to check that `key.modulus_size() >= 3072`?

**Comment by tarcieri** [comment]
I made an attempt at this in #526 but it's breaking the majority of current tests

**Comment by pinkforest** [comment]
Fixed them (by increasing key sizes) and I don't see tests taking too long as @ryancdotorg asserted correctly.

**Comment by tarcieri** [comment]
Thought I got this merged but ran into some test failures so I reverted.

I think a more incremental approach than #526, that first starts with placing limits on keygen, and then applies the limits generally to any loaded key (with escape hatches) might be in order.
